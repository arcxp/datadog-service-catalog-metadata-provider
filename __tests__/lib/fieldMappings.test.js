/**
 * @fileoverview This test covers all of the field mappings across versions.
 **/

const core = require('@actions/core')
const { mappings, mapField } = require('../../lib/fieldMappings')

describe.each([
  {
    version: 'v2',
    field: 'schema-version',
    input: 'v2',
    value: { 'schema-version': 'v2' },
  },
  {
    version: 'v2.1',
    field: 'schema-version',
    input: 'v2.1',
    value: { 'schema-version': 'v2.1' },
  },
  {
    version: 'v2',
    field: 'service-name',
    input: 'test-service',
    value: { 'dd-service': 'test-service' },
  },
  {
    version: 'v2.1',
    field: 'service-name',
    input: 'test-service',
    value: { 'dd-service': 'test-service' },
  },
  {
    version: 'v2',
    field: 'team',
    input: 'app-team-name',
    value: { team: 'app-team-name' },
  },
  {
    version: 'v2.1',
    field: 'team',
    input: 'app-team-name',
    value: { team: 'app-team-name' },
  },
  {
    version: 'v2',
    field: 'contacts',
    input: `
- name: Testy McTester
  type: email
  contact: testy@mctester.com
`,
    value: {
      contacts: [
        {
          name: 'Testy McTester',
          type: 'email',
          contact: 'testy@mctester.com',
        },
      ],
    },
  },
  {
    version: 'v2.1',
    field: 'contacts',
    input: `
- name: Testy McTester
  type: email
  contact: testy@mctester.com
`,
    value: {
      contacts: [
        {
          name: 'Testy McTester',
          type: 'email',
          contact: 'testy@mctester.com',
        },
      ],
    },
  },
  {
    version: 'v2',
    field: 'tags',
    input: `
- plain:nospaces
- compat1: space-in-value
- compat2 : space-in-both
`,
    value: {
      tags: [
        'plain:nospaces',
        'compat1:space-in-value',
        'compat2:space-in-both',
      ],
    },
  },
  {
    version: 'v2.1',
    field: 'tags',
    input: `
- plain:nospaces
- compat1: space-in-value
- compat2 : space-in-both
`,
    value: {
      tags: [
        'plain:nospaces',
        'compat1:space-in-value',
        'compat2:space-in-both',
      ],
    },
  },
  {
    version: 'v2',
    field: 'links',
    input: `
- name: 'first-link'
  type: url
  url: https://manchicken.io/testing
`,
    value: {
      links: [
        {
          name: 'first-link',
          type: 'url',
          url: 'https://manchicken.io/testing',
        },
      ],
    },
  },
  {
    version: 'v2',
    field: 'links',
    input: `
- name: 'first-link'
  type: url
  url: https://manchicken.io/testing
  provider: Website
`,
    value: {
      links: [
        {
          name: 'first-link',
          type: 'url',
          url: 'https://manchicken.io/testing',
        },
      ],
    },
  },
  {
    version: 'v2.1',
    field: 'links',
    input: `
- name: 'first-link'
  type: url
  url: https://manchicken.io/testing
  provider: Website
`,
    value: {
      links: [
        {
          name: 'first-link',
          type: 'url',
          url: 'https://manchicken.io/testing',
          provider: 'Website',
        },
      ],
    },
  },
  {
    version: 'v2',
    field: 'integrations',
    input: `
pagerduty: "https://my-org.pagerduty.com/service-directory/PMyService"
`,
    value: {
      integrations: {
        pagerduty: 'https://my-org.pagerduty.com/service-directory/PMyService',
      },
    },
  },
  {
    version: 'v2',
    field: 'integrations',
    input: `
opsgenie:
  service-url: "https://www.opsgenies.com/service/123e4567-e89b-12d3-a456-426614174000"
  region: US
`,
    value: {
      integrations: {
        opsgenie: {
          'service-url':
            'https://www.opsgenies.com/service/123e4567-e89b-12d3-a456-426614174000',
          region: 'US',
        },
      },
    },
  },
  {
    version: 'v2.1',
    field: 'integrations',
    input: `
pagerduty:
  service-url: "https://my-org.pagerduty.com/service-directory/PMyService"
`,
    value: {
      integrations: {
        pagerduty: {
          'service-url':
            'https://my-org.pagerduty.com/service-directory/PMyService',
        },
      },
    },
  },
  {
    version: 'v2.1',
    field: 'integrations',
    input: `
opsgenie:
  service-url: "https://www.opsgenies.com/service/123e4567-e89b-12d3-a456-426614174000"
  region: US
`,
    value: {
      integrations: {
        opsgenie: {
          'service-url':
            'https://www.opsgenies.com/service/123e4567-e89b-12d3-a456-426614174000',
          region: 'US',
        },
      },
    },
  },
  {
    version: 'v2',
    field: 'application',
    input: 'app-name',
    value: {
      error:
        'Sorry, but the «application» field is not avaiable in version v2 of the Datadog Service Catalog schema; this field is only available in version(s): v2.1',
    },
  },
  {
    version: 'v2.1',
    field: 'application',
    input: 'app-name',
    value: { application: 'app-name' },
  },
])('$field:$version', ({ version, field, input, value }) => {
  beforeEach(() => {
    core.setFailed.mockClear()
  })
  test('mapping', () => {
    if (value?.error) {
      expect(mapField(field, version)(input)).toBeUndefined()
      return
    }
    console.debug(
      JSON.stringify(
        { result: mapField(field, version)(input), value },
        undefined,
        2,
      ),
    )
    expect(mapField(field, version)(input)).toEqual(value)
  })

  test('failures', () => {
    if (!value?.error) {
      expect(core.setFailed).not.toHaveBeenCalled()
      return
    }

    console.debug(`Field: ${field} - Version: ${version}`, mappings.application)
    mapField(field, version)(input)
    expect(core.setFailed).toHaveBeenCalledTimes(1)
    expect(core.setFailed).toHaveBeenLastCalledWith(value.error)
  })
})
