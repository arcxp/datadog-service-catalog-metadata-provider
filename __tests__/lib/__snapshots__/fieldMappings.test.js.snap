// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constants convenienceFields 1`] = `
[
  "opsgenie",
  "pagerduty",
  "email",
  "slack-support-channel",
  "repo",
]
`;

exports[`constants mappings 1`] = `
{
  "application": {
    "v2": [Function],
    "v2.1": [Function],
  },
  "contacts": {
    "v2": [Function],
    "v2.1": [Function],
  },
  "description": {
    "v2": [Function],
    "v2.1": [Function],
  },
  "docs": {
    "v2": [Function],
    "v2.1": [Function],
  },
  "email": {
    "v2": [Function],
    "v2.1": [Function],
  },
  "integrations": {
    "v2": [Function],
    "v2.1": [Function],
  },
  "lifecycle": {
    "v2": [Function],
    "v2.1": [Function],
  },
  "links": {
    "v2": [Function],
    "v2.1": [Function],
  },
  "opsgenie": {
    "v2": [Function],
    "v2.1": [Function],
  },
  "pagerduty": {
    "v2": [Function],
    "v2.1": [Function],
  },
  "repo": {
    "v2": [Function],
    "v2.1": [Function],
  },
  "repos": {
    "v2": [Function],
    "v2.1": [Function],
  },
  "schema-version": {
    "v2": [Function],
    "v2.1": [Function],
  },
  "service-name": {
    "v2": [Function],
    "v2.1": [Function],
  },
  "slack-support-channel": {
    "v2": [Function],
    "v2.1": [Function],
  },
  "tags": {
    "v2": [Function],
    "v2.1": [Function],
  },
  "team": {
    "v2": [Function],
    "v2.1": [Function],
  },
  "tier": {
    "v2": [Function],
    "v2.1": [Function],
  },
}
`;

exports[`constants schemaFields 1`] = `
{
  "0": "schema-version",
  "1": "service-name",
  "10": "description",
  "11": "tier",
  "12": "lifecycle",
  "13": "email",
  "14": "slack-support-channel",
  "15": "repo",
  "16": "opsgenie",
  "17": "pagerduty",
  "2": "team",
  "3": "contacts",
  "4": "tags",
  "5": "links",
  "6": "integrations",
  "7": "docs",
  "8": "repos",
  "9": "application",
}
`;
